### DISCUSSION TITLE
Use environment variables for `Project Settings`

### DISCUSSION THREAD

**Discussion by ghost** [directus/directus#17154] [Feature Requests]
Imagine we have an `.env` file or environmental variables for our `docker-compose.yml` file, where we can specify parameters like `PROJECT_URL` or `DEFAULT_LANGUAGE` and if they are set, directus will disable those fields in the `project settings` and use them from the .env file.

**Comment by paescuj** [comment]
What advantages do you see compared to configuring the settings via API? Are you specifically interested in making them read-only?

**Reply by ghost** [reply]
@paescuj If we develop on localhost we have different values for our environmental variables than we have in production. In case of the `PROJECT_URL` it is something like `http://localhost:8055` while developing on a local machine and when deploying it will be something like `https://www.example.com`.

The advantage would be that we would separate environmental configs from app configs.

Maybe it is all about the `PROJECT_URL` …

**Reply by paescuj** [reply]
Okay, I see! Thank you for your answer!

**Reply by ernst86913** [reply]
You can actually define them in your .env… and then have an init hook which is passed the env and which updates the project settings…

**Reply by paescuj** [reply]
@ernst86913 Interesting approach 👍

**Reply by ghost** [reply]
Which init hook would I take? `init("app.before", () => {});`? How do I get access to the api to update project settings? Is there something like `this.api` ready for use?

**Reply by paescuj** [reply]
You can access internal services inside hooks, so you don't need to use the API for modifications.
A hook could look something like this (not tested!):
```js
export default ({ action }, { services, getSchema }) => {
	const { SettingsService } = services;

	action('server.start', async (_server, context) => {
		const settingsService = new SettingsService({ ...context, schema: await getSchema() });
		await settingsService.upsertSingleton({ project_url: process.env.PROJECT_URL })
	});
};
```

Another approach might be to intercept the read call on `settings` and return the modified data based on env vars. See https://github.com/directus/directus/discussions/16095#discussioncomment-3930543 for an example.

**Reply by ghost** [reply]
@paescuj Thanks, your code works great! Exactly what I’ve needed. The only thing I am wondering about, is how a developer could come to this solution. Is there any documentation about `services`, `getSchema`, … other than inspecting the code itself? It also doesn’t seem to be typed (?) — which could help. Of course, I’ve read [this page](https://docs.directus.io/extensions/hooks.html#action). So how to dig in deeper?

**Reply by paescuj** [reply]
The best source is indeed the official documentation. Context properties like `services` or `getSchema` are mentioned there.
The documentation is constantly being expanded and improved! In particular, a revision of the documentation on custom extension services is planned, see https://github.com/directus/docs/issues/20.
For now, you can rely on other resources like examples in [GitHub discussion](https://github.com/directus/directus/discussions) or the `#help` channel on our [Discord server](https://directus.chat/). And as always, the approach of learning by doing also applies here!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
